### ISSUE TITLE
Ingress controller deletes resources during control plane outage

### ISSUE THREAD

**Issue by jault3** [Kong/kubernetes-ingress-controller#4207]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Hello, I am running Kong Gateway and Kubernetes Ingress Controller (KIC) in a standard GKE cluster. GCP does periodic maintenance on the control plane (out of my hands) which causes the Kubernetes API server to be unavailable for some length of time (usually 5 minutes or so, but occasionally a bit longer). When this happens, the KIC container crashes and restarts. However, when it comes back up, it appears to permanently lose its connection to the k8s API server.

Once the API server finally becomes available again, one of the KIC containers grabs the leader lease and begins the sync process. Unfortunately, none of the ingress or service resources are found, and KIC ends up deleting every single resource in the gateway admin API (this includes all routes, services, plugins, ACLs, secrets, etc.). The gateway containers see this, process the changes, and all routing is gone. This issue is further exacerbated by the fact that it is unable to recover on its own without redeploying the pods.

### Expected Behavior

When the control plane is unavailable or any errors are encountered listing resources in the Kubernetes API, the ingress controller should not proceed with syncing. (Or at the very least, provide some sort of safeguard that prevents deleting all resources.)

### Steps To Reproduce

```markdown
I unfortunately do not have a reliable way of causing a control plane outage in a managed GKE cluster. However, I was able to simulate the behavior with these steps:

1. Deploy Kong and get it fully functional with routing working
2. Delete all cluster role bindings with `kubectl delete clusterrolebinding kong-ingress && kubectl delete clusterrolebinding kong-ingress-crds && kubectl delete clusterrolebinding kong-ingress-gateway && kubectl delete clusterrolebinding kong-ingress-knative`
3. Restart the pods with `kubectl rollout restart deploy ingress-kong`. Kong should come back up with empty caches, but fail to start any watches or get resources from the Kubernetes API server
4. Kong will grab the leader lease, and perform a sync operation. Once it finishes, all routes will be deleted

The only difference between this scenario and the real one I experience with GCP triggered control plane maintenance, is that when the cluster role bindings are restored, things begin working again. I haven't figured out why the control plane outage would cause it to be unable to recover on its own.

I tested this behavior on KIC 2.10.0, 2.9.3, 2.8.2, and 2.7.0 with the same result on each version.
```

### Kong Ingress Controller version

```shell
v2.10.0
```

### Kubernetes version

```shell
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.2", GitCommit:"7f6f68fdabc4df88cfea2dcf9a19b2b830f1e647", GitTreeState:"clean", BuildDate:"2023-05-17T14:13:27Z", GoVersion:"go1.20.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.8-gke.1000", GitCommit:"4ac1389d4c3eefaabbd9fa31782fcbfd72e6e6e6", GitTreeState:"clean", BuildDate:"2023-03-30T14:01:58Z", GoVersion:"go1.19.7 X:boringcrypto", Compiler:"gc", Platform:"linux/amd64"}

GKE control plane is running version `1.25.8-gke.1000`
```

### Anything else?

* I am running OSS gateway version 3.3.0 in traditional mode with a postgres database
* Prior to this, I was running KIC 2.4.2 and gateway 2.8.3 and it did _not_ exhibit this behavior.
* I have attached the debug logs from the KIC container when I simulated this using the steps above. Resource names have been redacted from the log.
* I am running the YAML resources defined here https://github.com/Kong/kubernetes-ingress-controller/blob/v2.10.0/deploy/single/all-in-one-postgres.yaml with some modifications for my environment, but the KIC container definition is the same.

[ingress-controller-logs.txt](https://github.com/Kong/kubernetes-ingress-controller/files/11826649/ingress-controller-logs.txt)

**Comment by rainest** [comment]
tl;dr hard to confirm, but this may be caused by some logic we use to determine whether to run controllers for individual resources being unable to see resource definitions at start, so it disables those controllers and cannot see the resources behind the configuration it's deleting. I'm unsure what would make sense as a fix.

---

Do you see the same "cannot list" errors that you get when deleting roles during an actual API server outage? It looks like the attached logs are from the former, and I'm more interested in the latter. Dunno if you have historical logs around from the last outage.

[KIND](https://kind.sigs.k8s.io/) should let you replicate closer to the actual scenario. My rough steps to simulate the issue were:
- Deploy an instance, `docker exec -it CONTAINER_ID /bin/bash` into it
- `apt update; apt install vi`, `vi /etc/kubernetes/manifests/kube-apiserver.yaml`, and replace the `kube-apiserver` in the command to a garbage string
- Save the file, wait a bit and reverted the change.

KIND will attempt to deploy that file's contents as the API server Pod, so if it's broken the API server will be offline. The controller loses the lease and terminates shortly after:

```
E0626 19:37:45.159943       1 leaderelection.go:327] error retrieving resource lock kong/5b374a9e.konghq.com: Get "https://10.96.0.1:443/apis/coordination.k8s.io/v1/namespaces/kong/leases/5b374a9e.konghq.com": dial tcp 10.96.0.1:443: connect: connection refused
I0626 19:37:47.158545       1 leaderelection.go:280] failed to renew lease kong/5b374a9e.konghq.com: timed out waiting for the condition
Error: leader election lost
```

I don't think we can avoid that since the leader election behavior is [implemented by an upstream library](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/leaderelection). At most we'd be able to extend the lease grace period (maybe--it doesn't look like controller-runtime lets you configure most of the underlying client-go leadership library settings).

AFAIK the controller is successfully connecting to the API server after it restarts. Leader election is handled via a ConfigMap, so it can't read or update that and check/acquire the lease without the API server.

When my instance comes back online, I see several sections of interest in the logs:

```
time="2023-06-26T19:55:06Z" level=info msg="Disabling controller for Group=configuration.konghq.com/v1beta1, Resource=udpingresses due to missing CRD" logger=controllers.crdCondition
...
I0626 19:55:06.297828       1 leaderelection.go:245] attempting to acquire leader lease kong/5b374a9e.konghq.com...
E0626 19:55:06.298020       1 leaderelection.go:327] error retrieving resource lock kong/5b374a9e.konghq.com: Get "https://10.96.0.1:443/apis/coordination.k8s.io/v1/namespaces/kong/leases/5b374a9e.konghq.com": dial tcp 10.96.0.1:443: connect: connection refused
...
I0626 19:57:10.367582       1 leaderelection.go:255] successfully acquired lease kong/5b374a9e.konghq.com
time="2023-06-26T19:57:10Z" level=info msg="ingress-kong-7c96dccb48-fmk2n_586e64a1-3f16-4407-b3b9-744ad62605b0 became leader" logger=events object="{\"kind\":\"Lease\",\"namespace\":\"kong\",\"name\":\"5b374a9e.konghq.com\",\"uid\":\"90b4532c-50aa-460f-b8af-a9aea2cde9a0\",\"apiVersion\":\"coordination.k8s.io/v1\",\"resourceVersion\":\"21414\"}" reason=LeaderElection type=Normal
...
time="2023-06-26T19:57:10Z" level=info msg="Starting EventSource" logger=controllers.EndpointSlice source="kind source: *v1.EndpointSlice"
time="2023-06-26T19:57:10Z" level=info msg="Starting EventSource" logger=controllers.Secrets source="kind source: *v1.Secret"
time="2023-06-26T19:57:10Z" level=info msg="Starting EventSource" logger=controllers.Dynamic/Gateway source="kind source: *v1.CustomResourceDefinition"
time="2023-06-26T19:57:10Z" level=info msg="Starting EventSource" logger=controllers.Dynamic/HTTPRoute source="kind source: *v1.CustomResourceDefinition"
time="2023-06-26T19:57:10Z" level=info msg="Starting EventSource" logger=controllers.Service source="kind source: *v1.Service"
...
deleting plugin correlation-id for route default.httpbin.httpbin..80
deleting route default.httpbin.httpbin..80
deleting target 10.244.0.13:80 for upstream a5c40eb3-77ac-43ec-af41-51aeab32924a
deleting service default.httpbin.httpbin.80
deleting upstream httpbin.default.80.svc
```

So it starts, initially fails to grab the lease, connects and grabs it, starts resource controllers, and then deletes configuration. The controller disable at start and EventSource (resource controller) starts look like a possible cause. Those are the _only_ resource controllers started, whereas normally you'd see some for Ingress and such also. If I restart the controller when the API server is fully online, I get:

```
time="2023-06-26T20:01:57Z" level=info msg="Starting EventSource" logger=controllers.Ingress.netv1 source="channel source: 0xc000561400"
time="2023-06-26T20:01:57Z" level=info msg="Starting EventSource" logger=controllers.Ingress.netv1 source="kind source: *v1.IngressClass"
```
along with the others, and the controller creates the route it had deleted.

Current versions of the controller ([starting with 2.8](https://github.com/Kong/kubernetes-ingress-controller/pull/3013)) revised resource controller loading to disable controllers when their resource definitions are missing. These checks fail if the API server is offline, and only run once at startup.

I'm not sure how this should affect the Ingress resource controller, however, as it has used a similar check to determine which API version it should use [since 2.0](https://github.com/Kong/kubernetes-ingress-controller/commit/0c43235e4762da78ca09d7d7a82123ebc4ea00fa). I'd expect that to have issues even on 2.4.

On 2.7 my replication steps do not reproduce the issue, but it's not clear that this is due to the earlier version of the resource controller setup. I cannot seem to get it into the state where it starts with the API server offline and manages to grab the lease after. It always grabs the lease on the first try.

**Comment by jault3** [comment]
I don't have debug logs available for an actual outage, but we did have another one this morning and I have the normal info level logs from it: [logs-pt2.csv](https://github.com/Kong/kubernetes-ingress-controller/files/11918883/logs-pt2.csv). I do not see the "cannot list resources" logs though.

But yeah I agree, losing the leader election during an outage like this is expected. It does seem like the "disabling controller" logs are the cause. I would think ingresses would be a required resource to be able to query for on startup, so it was a bit strange that it proceeds without trying again after an error. Not sure what a good fix for it would be, but maybe if it can't reach the control plane, it shouldn't do any startup processes at all? or delay the resource existence checks (seeing what CRDs are available) until after its grabbed the lease?

**Comment by jault3** [comment]
Hi @rainest - just wanted to check in and see if there was any other info you needed from me, or if there was any progress on a fix for this?

**Comment by rainest** [comment]
We can replicate something similar enough independently and would not need further information from user environments.

I don't have any additional updates on scheduling work to fix this at this time.

**Comment by gallolp** [comment]
Hi, just chiming in to say I am experiencing this issue with KIC 2.10.x but not with 2.7.x.

In my case I have a single release/one pod deployment (proxy+ingress on the same pod). On k8s cluster startup a short network outage makes the KIC unable to discover k8s api groups. Logging:
```
failed to get API group resources: unable to retrieve the complete list of server APIs: v1: Get "https://172.20.0.1:443/api/v1": dial tcp 172.20.0.1:443: connect: connection refused
```
and
```
Disabling controller for Group=configuration.konghq.com/v1, Resource=kongplugins due to missing CRD
```
for each of the enabled resources.

The ingress-controller never retries to discover k8s api groups.

**Comment by rainest** [comment]
I found new ways to interact with a broken cluster and [took notes from poking at it more](https://gist.github.com/rainest/5e651b2d15bcb2a9931373750b6efdc3).

Notable takeaway is that offline API servers generate no events that would update the cached store, so the controller can run with it offline and will continue pushing whatever configuration it last had. We have not exposed the leadership election lease timeout, however, so you can't force it to continue running longer than the outage duration. Changing this would also have some other implications for recovery after a controller crash.

I'd originally considered allowing controllers to fail in controller-runtime and/or disabling resource kind detection, but these don't help much if the server remains inaccessible beyond whenever config generation starts. I think we simply [have to fail immediately](https://github.com/Kong/kubernetes-ingress-controller/pull/4626) if we can't reach it and allow the Deployment to manage retries.

**Comment by rainest** [comment]
Github was picky about language and didn't auto-close, but #4641 is expected to avoid this (the controller won't start if the API server isn't available) and will be out in 2.12 in the next week or so.

**Comment by pmalek** [comment]
@jault3 You can also start experimenting with this using our nightly images https://hub.docker.com/r/kong/nightly-ingress-controller/tags

**Comment by backjo** [comment]
Seeing a somewhat related issue to this when the controller is not configured correctly - we had a case where we configured the feature gate for GRPCRoute and other CRDs but did not have the role correct for the controller. Because the controller, could not read those resource, the controller didn't build a Kong configuration but ended up posting an empty configuration to the data plane.
